From dfaa29de7fd7b0b1344cc25566436ab0e4d38f7f Mon Sep 17 00:00:00 2001
From: whitequark <whitequark@whitequark.org>
Date: Wed, 17 Aug 2016 11:00:50 +0000
Subject: [PATCH] [OR1K] Follow placement of _GLOBAL_OFFSET_TABLE_ in
 R_OR1K_GOTOFF_*.

_bfd_elf_create_got_section places the _GLOBAL_OFFSET_TABLE_ symbol
either at the start of .got.plt (if one is placed) or .got (if not).
However, R_OR1K_GOTOFF_* are always computed relative to .got, so
if a PLT is generated, the offsets are incorrect.

This commit alters the relocation logic so that it follows the logic
in _bfd_elf_create_got_section.
---
 bfd/elf32-or1k.c | 15 +++++++++++----
 1 file changed, 11 insertions(+), 4 deletions(-)

diff --git a/bfd/elf32-or1k.c b/bfd/elf32-or1k.c
index 1daa7b6..522f141 100644
--- a/bfd/elf32-or1k.c
+++ b/bfd/elf32-or1k.c
@@ -809,7 +809,7 @@ or1k_elf_relocate_section (bfd *output_bfd,
   bfd *dynobj;
   asection *sreloc;
   bfd_vma *local_got_offsets;
-  asection *sgot;
+  asection *sgot, *sgotplt;
 
   if (htab == NULL)
     return FALSE;
@@ -820,6 +820,7 @@ or1k_elf_relocate_section (bfd *output_bfd,
   sreloc = elf_section_data (input_section)->sreloc;
 
   sgot = htab->sgot;
+  sgotplt = htab->sgotplt;
 
   symtab_hdr = &elf_tdata (input_bfd)->symtab_hdr;
   sym_hashes = elf_sym_hashes (input_bfd);
@@ -996,9 +997,15 @@ or1k_elf_relocate_section (bfd *output_bfd,
 
         case R_OR1K_GOTOFF_LO16:
         case R_OR1K_GOTOFF_HI16:
-          /* Relocation is offset from GOT.  */
-          BFD_ASSERT (sgot != NULL);
-          relocation -= sgot->output_section->vma;
+          /* Relocation is offset from the _GLOBAL_OFFSET_TABLE_ symbol, which is
+             located at the start of PLT (if there's one) or GOT.
+             See _bfd_elf_create_got_section.  */
+          if(sgotplt != NULL) {
+            relocation -= sgotplt->output_section->vma;
+          } else {
+            BFD_ASSERT (sgot != NULL);
+            relocation -= sgot->output_section->vma;
+          }
           break;
 
         case R_OR1K_INSN_REL_26:
-- 
2.8.1

